### ISSUE TITLE
Add ability to inject CommandLine into Commands

### ISSUE THREAD

**Issue by bobtiernay-okta** [remkop/picocli#368]
This would be super handy for implementing commands such as `git`. The parent command `git` simply prints this:

```
usage: git [--version] [--help] [-C <path>] [-c name=value]
           [--exec-path[=<path>]] [--html-path] [--man-path] [--info-path]
           [-p | --paginate | --no-pager] [--no-replace-objects] [--bare]
           [--git-dir=<path>] [--work-tree=<path>] [--namespace=<name>]
           <command> [<args>]

These are common Git commands used in various situations:

start a working area (see also: git help tutorial)
   clone      Clone a repository into a new directory
   init       Create an empty Git repository or reinitialize an existing one

work on the current change (see also: git help everyday)
   add        Add file contents to the index
   mv         Move or rename a file, a directory, or a symlink
   reset      Reset current HEAD to the specified state
   rm         Remove files from the working tree and from the index

examine the history and state (see also: git help revisions)
   bisect     Use binary search to find the commit that introduced a bug
   grep       Print lines matching a pattern
   log        Show commit logs
   show       Show various types of objects
   status     Show the working tree status

grow, mark and tweak your common history
   branch     List, create, or delete branches
   checkout   Switch branches or restore working tree files
   commit     Record changes to the repository
   diff       Show changes between commits, commit and working tree, etc
   merge      Join two or more development histories together
   rebase     Reapply commits on top of another base tip
   tag        Create, list, delete or verify a tag object signed with GPG

collaborate (see also: git help workflows)
   fetch      Download objects and refs from another repository
   pull       Fetch from and integrate with another repository or a local branch
   push       Update remote refs along with associated objects

'git help -a' and 'git help -g' list available subcommands and some
concept guides. See 'git help <command>' or 'git help <concept>'
to read about a specific subcommand or concept.
```

If there are additional configurations performed in `CommandLine`, then these need to be reconstructed in the parent command. Allowing for injection of the `CommandLine` instance would help here.

**Comment by remkop** [comment]
Can you clarify your use case a bit? Picocli currently provides
* `@ParentCommand` [annotation](http://picocli.info/#__code_parentcommand_code_annotation)
* the `IHelpCommandInitializable` interface for [custom help commands](http://picocli.info/#_custom_help_subcommands)

The latter interface can provide help commands with their `CommandLine` object so they can discover details about their parent and sibling commands in order to create a good help message. If your use case is a custom help subcommand then I recommend marking the command as `helpCommand = true` and implementing this interface.

I am not opposed to adding an `@Inject` annotation to inject a `CommandLine` instance but I would like to understand why the above are insufficient for your use case. :-)

**Comment by remkop** [comment]
Can you clarify if this is for a custom `help` command or a different use case?

**Comment by bobtiernay-okta** [comment]
Originally that was the idea, but I could see this being generally useful to be able to introspect the command line object.

**Comment by bobtiernay-okta** [comment]
Please do, thanks!

**Comment by remkop** [comment]
FYI: I ran into a use case that requires access to the picocli command model from the application, so this will be available from picocli 3.2.

The use case has to do with JLine completion candidates (#393, also in v3.2).
Some completion candidates need to be looked up dynamically from some resource that is only available at runtime. Access to the `CommandSpec` is needed to update the option's `ArgSpec.completionCandidates` after the lookup.
